### ISSUE TITLE
MacOS: fails to configure correctly - insists on cross-compile

### ISSUE THREAD

**Issue by mouse07410** [objectcomputing/OpenDDS#1479]
Current master. MacOS Mojave 10.14.6, Xcode-11.3.1.

Build script:
```bash
#!/bin/bash -x

make clean || true

if [  -z "${CXX}" ]; then
    CXX="clang++"
fi

CXXFLAGS="-Os -Ofast -march=native -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk"

OPENDDS_COMPILER="$CXX"

export QTDIR=/opt/local/libexec/qt5
export DDS_ROOT="${PWD}"
DAT=`date "+%Y-%m-%d"`

echo "########################################"
echo "# Building OpenDDS                     #"
echo "DDS_ROOT=\"${DDS_ROOT}\""
echo "CXX=\"${CXX}\""
echo "########################################"

./configure  --verbose \
             --prefix="${HOME}/OpenDDS" \
             --std=c++11 \
             --java \
             --xerces3=/opt/local \
             --qt=/opt/local/libexec/qt5 \
             --safety-profile=base \
             --security \
             --openssl=/opt/local \
             --rapidjson \
             --gtest=/opt/local \
             --tests \
             --cmake=/opt/local/bin/cmake \
             --ace=${PWD}/ACE_TAO/ACE \
             --tao=${PWD}/ACE_TAO/TAO \
             --mpc=${PWD}/MPC \
             --ace-github-latest \
             --boost=/opt/local  2>&1 | tee conf-out-${DAT}.txt

time make -j 4 all 2>&1 | tee make-out.txt
if [ $? != 0 ]; then
	echo "OpenDDS build failed!"
	exit 1
fi

. setenv.sh
time bin/auto_run_tests.pl 2>&1 | tee tests-out-${DAT}.txt

set +x
sleep 1
echo "Tests PASSED: " `fgrep "test PASSED" tests-out-${DAT}.txt | wc -l`
echo "Tests FAILED: " `fgrep "test FAILED" tests-out-${DAT}.txt | wc -l`
```

Here's what the config says:
```
Options:
'ace' => '/Users/ur20980/src/OpenDDS/ACE_TAO/ACE'
'ace-github-latest' => 1
'boost' => '/opt/local'
'cmake' => '/opt/local/bin/cmake'
'gtest' => '/opt/local'
'java' => ''
'mpc' => '/Users/ur20980/src/OpenDDS/MPC'
'mpcopts' => ARRAY(0x7fbd23016110)
   0  '-workers 4'
'openssl' => '/opt/local'
'prefix' => '/Users/ur20980/OpenDDS'
'qt' => '/opt/local/libexec/qt5'
'rapidjson' => '/opt/local'
'safety-profile' => 'base'
'security' => 1
'std' => 'c++11'
'tao' => '/Users/ur20980/src/OpenDDS/ACE_TAO/TAO'
'tests' => 1
'verbose' => 1
'xerces3' => '/opt/local'
host system is: macosx
Cross-compile build including host tools
Auto-detecting compiler
Found clang++ at: /usr/bin/clang++
compiler is: clang++
Added platform_macros for -std=c++11
Compiler has C++11 support
Using ace_src: /Users/ur20980/src/OpenDDS/ACE_TAO/ACE
Using tao_src: /Users/ur20980/src/OpenDDS/ACE_TAO/TAO
Using CMake '/opt/local/bin/cmake' if needed
javac 1.8.0_242
Enabling xerces3
Enabling boost
Enabling java
Enabling rapidjson
Enabling openssl
Enabling qt
cloning build tree
Running "/Users/ur20980/src/OpenDDS/MPC/clone_build_tree.pl host target"
Can't exec "/Users/ur20980/src/OpenDDS/MPC/clone_build_tree.pl": No such file or directory at ./configure line 408.
Command "/Users/ur20980/src/OpenDDS/MPC/clone_build_tree.pl host target" Failed to Run: No such file or directory
ERROR: Can't write to build/host/host_tools.mwc, stopped at ./configure line 925.
Cross-compile configuring host
```

Why does it insist now on cross-compiling, and how to modify the `./configure` parameters to avoid it? What am I missing, or what is extra?

**Comment by iguessthislldo** [comment]
I got to say that's a lot of options to pass to the configure script. It's almost cruel to the configure script...

Anyway the specific cause of this error is that you gave it a bogus `--mpc` option. The `MPC` folder in the OpenDDS source directory is where we store config files for MPC, but not MPC itself. You should let configure script handle MPC, as well as ACE and TAO unless you know you want to use a custom one. This also means that the `--ace` and `--tao` options are unnecessary for the same reason, but in this case they're harmless because that's the locations where `--ace-github-latest` puts them.

As for the cross compilation part, you passed `--safety-profile=base`. Safety profile is a special memory restricted mode that is built using something very similar to cross compilation, so it uses. I'm not completely familiar with other features it works with, but I definitely wouldn't recommend it unless you **absolutely know** you need it.

Some other notes:
- Unlike some other make files, `CXXFLAGS` isn't going to be used by our build system. ACE Makefiles uses `CPPFLAGS` for this purpose, but I'm not sure if ACE supports setting an initial value for it from the environment or not. See if you get what you want without setting something like that first.
- I would also not recommend providing a custom Google Test. OpenDDS uses a specific commit of Google Test and the configure script will build it automatically if it has cmake, which you are giving it.
- This option: `--prefix="${HOME}/OpenDDS"` seems a little odd to me. This is where OpenDDS would get installed to if you ran `make install`. If that's what you want, then this options is okay.
- ` --boost=/opt/local` is only used in the ishapes demo if C++11 isn't available.

In summary, I would guess you don't really need to configure OpenDDS with all of these options, but if you really want to, this is what should work given the information you've given:

```
./configure  --verbose \
             --std=c++11 \
             --java \
             --xerces3=/opt/local \
             --qt=/opt/local/libexec/qt5 \
             --security \
             --openssl=/opt/local \
             --rapidjson \
             --tests \
             --cmake=/opt/local/bin/cmake \
             --ace-github-latest 2>&1 | tee conf-out-${DAT}.txt
```

**Comment by mouse07410** [comment]
@iguessthislldo thank you very much! You provided the info I couldn't find. Incorporating your recommendations, updated build script is:
```bash
#!/bin/bash -x

make clean || true

if [  -z "${CXX}" ]; then
    CXX="clang++"
fi

CPPFLAGS="-Os -Ofast -march=native -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk"

OPENDDS_COMPILER="$CXX"

OPENDDS_INSTALL_DIR="${HOME}/OpenDDS"

export QTDIR=/opt/local/libexec/qt5
export DDS_ROOT="${PWD}"
DAT=`date "+%Y-%m-%d"`

echo "########################################"
echo "# Building OpenDDS                     #"
echo "DDS_ROOT=\"${DDS_ROOT}\""
echo "CXX=\"${CXX}\""
echo "########################################"

./configure  --verbose \
             --prefix="${OPENDDS_INSTALL_DIR}" \
             --std=c++11 \
             --java \
             --xerces3=/opt/local \
             --qt=${QTDIR} \
             --security \
             --openssl=/opt/local \
             --tests \
             --cmake=/opt/local/bin/cmake \
             --ace-github-latest 2>&1 | tee conf-out-${DAT}.txt

time make -j 4 all 2>&1 | tee make-out.txt
if [ $? != 0 ]; then
	echo "OpenDDS build failed!"
	exit 1
fi

. setenv.sh
time bin/auto_run_tests.pl 2>&1 | tee tests-out-${DAT}.txt

set +x
sleep 1
echo "Tests PASSED: " `fgrep "test PASSED" tests-out-${DAT}.txt | wc -l`
echo "Tests FAILED: " `fgrep "test FAILED" tests-out-${DAT}.txt | wc -l`
```

The build appears to proceed much better now, but the configurator still fails (though in a different place):
```
. . . . .
Skipping FaceUnitTests_SafetyProfilePool (FaceUnitTests.mpc); it avoids no_opendds_safety_profile.
Skipping dissector (dissector.mpc); it requires wireshark_any.
./MessengerSplit.mpc: line 1:
ERROR: Cyclic inheritance detected: MessengerSplit
MessengerSplit.mpc: line 1:
ERROR: Invalid parent: MessengerSplit
DDS.mwc: line 21:
ERROR: Unable to process build/target/tools/modeling/tests/MessengerSplit/MessengerSplit.mpc
ERROR: Unable to process: DDS.mwc
Command "perl /Users/ur20980/src/OpenDDS/ACE_TAO/ACE/bin/mwc.pl -type gnuace /Users/ur20980/src/OpenDDS/DDS.mwc -features xerces3=1 -features ssl=1 -features qt5=1 -features no_rapidjson=0 -features java=1 -features openssl11=1 -features no_opendds_security=0" Returned 1
ERROR: Error from MPC, stopped at ./configure line 1988.
```

Consequently, the build fails:
```
. . . . .
clang++ -std=c++11 -fdelayed-template-parsing -g -m64 -W -Wall -DACE_HAS_CUSTOM_EXPORT_MACROS=0   -I/Users/ur20980/src/OpenDDS/build/target/ACE_TAO/ACE -D__ACE_INLINE__ -I/Users/ur20980/src/OpenDDS/build/target/ACE_TAO/ACE  -m64 -flat_namespace -rpath @executable_path/. -rpath "/Users/ur20980/src/OpenDDS/build/target/ACE_TAO/ACE/lib" -L/Users/ur20980/src/OpenDDS/build/target/ACE_TAO/ACE/lib -L.  -L/Users/ur20980/src/OpenDDS/build/target/ACE_TAO/ACE/lib  -o stub .obj/stub.o -lACE  
Undefined symbols for architecture x86_64:
  "__ace_assert(char const*, int, char const*)", referenced from:
      CORBA::Exception::Exception(char const*, char const*) in Exception.o
      CORBA::Exception::Exception(CORBA::Exception const&) in Exception.o
      CORBA::Exception::operator=(CORBA::Exception const&) in Exception.o
  "ACE_Log_Msg::conditional_set(char const*, int, int, int)", referenced from:
      CORBA::Exception::_tao_print_exception(char const*, __sFILE*) const in Exception.o
      CORBA::SystemException::_tao_print_system_exception(__sFILE*) const in SystemException.o
  "ACE_Log_Msg::last_error_adapter()", referenced from:
      CORBA::Exception::_tao_print_exception(char const*, __sFILE*) const in Exception.o
      CORBA::SystemException::_tao_print_system_exception(__sFILE*) const in SystemException.o
  "ACE_Log_Msg::log(ACE_Log_Priority, char const*, ...)", referenced from:
      CORBA::Exception::_tao_print_exception(char const*, __sFILE*) const in Exception.o
      CORBA::SystemException::_tao_print_system_exception(__sFILE*) const in SystemException.o
  "ACE_Log_Msg::instance()", referenced from:
      CORBA::Exception::_tao_print_exception(char const*, __sFILE*) const in Exception.o
      CORBA::SystemException::_tao_print_system_exception(__sFILE*) const in SystemException.o
  "ACE_Allocator::instance()", referenced from:
      CORBA::string_alloc(unsigned int) in String_Alloc.o
      CORBA::string_free(char*) in String_Alloc.o
      CORBA::wstring_alloc(unsigned int) in String_Alloc.o
      CORBA::wstring_free(wchar_t*) in String_Alloc.o
      CORBA::UNKNOWN::_tao_create() in SystemException.o
      CORBA::BAD_PARAM::_tao_create() in SystemException.o
      CORBA::NO_MEMORY::_tao_create() in SystemException.o
      ...
  "ACE_Thread_Mutex::ACE_Thread_Mutex(char const*, _opaque_pthread_mutexattr_t*)", referenced from:
      CORBA::Object::Object(TAO_Stub*, bool, TAO_Abstract_ServantBase*, TAO_ORB_Core*) in Object.o
      CORBA::Object::Object(IOP::IOR*, TAO_ORB_Core*) in Object.o
  "ACE_Thread_Mutex::~ACE_Thread_Mutex()", referenced from:
      CORBA::Object::~Object() in Object.o
  "ACE_Message_Block::~ACE_Message_Block()", referenced from:
      ACE_InputCDR::~ACE_InputCDR() in Exception.o
      ACE_InputCDR::~ACE_InputCDR() in LocalObject.o
      ACE_InputCDR::~ACE_InputCDR() in Object.o
      ACE_InputCDR::~ACE_InputCDR() in SystemException.o
  "ACE_String_Base_Const::npos", referenced from:
      ACE_String_Base<char>::append(char const*, unsigned long) in SystemException.o
  "ACE_OS::snprintf(char*, unsigned long, char const*, ...)", referenced from:
      CORBA::SystemException::_info() const in SystemException.o
  "ACE_OS::strerror(int)", referenced from:
      CORBA::SystemException::_info() const in SystemException.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [libOpenDDS_Corba.dylib] Error 1
make[1]: *** [OpenDDS_Corba] Error 2
make[1]: *** Waiting for unfinished jobs....
make: *** [all] Error 2
```

**Comment by mouse07410** [comment]
Took the above script, cloned this repo afresh:
```
$ git clone https://github.com/objectcomputing/OpenDDS.git
. . . . .
$ cd OpenDDS
$ git submodule update --init --recursive
. . . . .
```

Then copied the above script, and ran the build. It appears to succeed: 258 tests passed, 11 failed.

Here's the complete output:
[conf-out-2020-01-25.txt](https://github.com/objectcomputing/OpenDDS/files/4112294/conf-out-2020-01-25.txt)
[make-out.txt](https://github.com/objectcomputing/OpenDDS/files/4112296/make-out.txt)
[tests-out-2020-01-25.txt](https://github.com/objectcomputing/OpenDDS/files/4112297/tests-out-2020-01-25.txt)

@iguessthislldo I intend to use OpenDDS with security enabled (having said that - I haven't yet studied what security capabilities OpenDDS offers, and how they compare to, e.g., TLS). If you have time to take a look at the failed tests above - could you offer any more advice?

Thank you!

**Comment by iguessthislldo** [comment]
> Then copied the above script, and ran the build. It appears to succeed: 258 tests passed, 11 failed.

Yeah the configure script isn't really meant to be ran multiple times, especially with widely different configurations. For future reference you can use `git clean -dfX` to clean a git repo. You would also have to do the same to `ACE_TAO`.

> If you have time to take a look at the failed tests above - could you offer any further advice?

There are some issues with macOS that could cause tests to fail, like the lack of support for the shared memory transport. For reference these are the `-Config` flags we pass to `auto_run_test.pl` on our macOS CI build: `-Config RTPS -Config NO_SHMEM -Config DDS_NO_ORBSVCS -Config CXX11` to tell it what tests to run. It's also worth noting that master branch isn't quite stable as it's not guaranteed to be be. You also seem particularly interested in running tests, so it's probably worth mentioning that there are separate test list files for Java `java/tests/dcps_java_tests.lst` and Security `tests/security/security_tests.lst` that won't be ran by default.

If you're serious about getting a specific problem with OpenDDS fixed, please either create a pull request or see our commercial support.

**Comment by mouse07410** [comment]
@iguessthislldo 

> you can use `git clean -dfX` to clean a git repo

Thank you! Will do.

> these are the `-Config` flags we pass to `auto_run_test.pl` on our macOS CI build...

Could you clarify please - are *all* of the flags you listed already/automatically passed to the test-run script? And if I invoke the script manually, can I pass more than one `-Config` flag?

> Security `tests/security/security_tests.lst`...

Could you please show how I'd invoke these tests? What command line to use?

Thank you!

> If you're serious about getting a specific problem with OpenDDS fixed...

First, I need to make sure the problem is with OpenDDS itself, rather than with what I'm trying to do with it. ;-)

**Comment by mitza-oci** [comment]
> @iguessthislldo thank you very much! You provided the info I couldn't find. Incorporating your recommendations

Please let us know if documentation updates are in order.  Safety profile has a chapter in the developer's guide.  I'm not sure which other items fell into the "I couldn't find" category.

**Comment by mitza-oci** [comment]
> > these are the `-Config` flags we pass to `auto_run_test.pl` on our macOS CI build...
> 
> Could you clarify please - are _all_ of the flags you listed already/automatically passed to the test-run script? And if I invoke the script manually, can I pass more than one `-Config` flag?
> 
> > Security `tests/security/security_tests.lst`...
> 
> Could you please show how I'd invoke these tests? What command line to use?

Start with `bin/auto_run_tests.pl --help`, hopefully that will make the notes from @iguessthislldo more clear

**Comment by mouse07410** [comment]
> ...hopefully that will make the notes ... more clear

I'm afraid I'm even more confused. 

```
$ bin/auto_run_tests.pl --help
Unknown option: help
auto_run_tests.pl [-a] [-h] [-s sandbox] [-o] [-t] [-l listfile]

Runs the tests listed in dcps_tests.lst

Options:
    -h          display this help
    -c config   Run the tests for the <config> configuration
    -Config cfg Run the tests for the <cfg> configuration
    -s sandbox  Runs each program using a sandbox program
    -a          Run all DDS (DCPS) tests (default unless -l)
    -x          Stop on any failure
    -l listfile Run the tests specified in list file

DCPS Test Configs: DCPS_MIN OPENDDS_SAFETY_PROFILE NO_MCAST RTPS DDS_NO_OWNERSHIP_PROFILE DDS_NO_OBJECT_MODEL_PROFILE DDS_NO_QUERY_CONDITION STATIC XERCES3 NO_BUILT_IN_TOPICS NO_SHMEM CXX11 RAPIDJSON DDS_NO_PERSISTENCE_PROFILE TARGET NO_DDS_TRANSPORT MIN_CORBA IPV6 DDS_NO_ORBSVCS CORBA_E_COMPACT CORBA_E_MICRO DDS_NO_CONTENT_SUBSCRIPTION DDS_NO_CONTENT_FILTERED_TOPIC DDS_NO_MULTI_TOPIC QNX LYNXOS DDS_NO_OWNERSHIP_KIND_EXCLUSIVE WCHAR
```

Does it mean that only one `-Config` flag can be passed at one invocation? According to the notes, I assume yes, according to `--help` - it seems only one can be passed?

What is `-o` option that wasn't described? What is `-t`? Why isn't `-x` listed on the command line help? Why doesn't `-a` (presumably *all* the tests) include Security tests?

`INSTALL.md` says:
```
Optionally, you can run the entire OpenDDS regression test suite with one Perl command.

**For Unixes (Linux, macOS, Solaris, BSDs, etc):**

bin/auto_run_tests.pl
```
Are Security tests considered a part of the entire OpenDDS regression test suite? If not - how do I invoke them? By adding `-l tests/security/security_tests.lst`? Anything else required...?

I saw no `-Config ...` present in the `.travis.yml` - how are you passing those flags to CI?

I'm very confused.

**Comment by iguessthislldo** [comment]
> I'm afraid I'm even more confused.

> What is -o option that wasn't described? What is -t? Why isn't -x listed on the command line help? Why doesn't -a (presumably all the tests) include Security tests?

Yeah, I had the same reaction. Something needs to be done about this script because I recently [cleaned up its counterpart in ACE](https://github.com/DOCGroup/ACE_TAO/pull/1014), which had the same sort of problems. It either needs to have the same treatment or maybe even be removed. I also would like to have some sort of include functionality so the extra test lists will also be run while keeping them as separate files.

As a workaround, see if you can use `ACE_TAO/ACE/bin/auto_run_tests.pl` instead. After sourcing `setenv.sh`, try something like:

```
$ACE_ROOT/bin/auto_run_tests.pl \
  -d \
  $DDS_ROOT/tests/security/security_tests.lst \
  $DDS_ROOT/java/tests/dcps_java_tests.lst \
  -Config RTPS -Config NO_SHMEM -Config DDS_NO_ORBSVCS -Config CXX11
```

`$ACE_ROOT/bin/auto_run_tests.pl` has a more helpful `--help` message too.

> Does it mean that only one -Config flag can be passed at one invocation? According to the notes, I assume yes, according to --help - it seems only one can be passed?

Multiple `-Config`s can be passed.

> I saw no -Config ... present in the .travis.yml - how are you passing those flags to CI?

We are time limited on CI like Travis, so we don't run the test suite, we just build them. We do run tests for member of the OpenDDS team on our own custom CI that does invoke auto_run_tests.

**Comment by mouse07410** [comment]
BTW, a naive question re. `NO_SHMEM`. Since MacOS is a Unix (BSD flavor, rather than Linux - but I ran a lot of Linux software on it), is there a reason SHMEM shouldn't work on Mac as is?

**Comment by iguessthislldo** [comment]
> is there a reason SHMEM shouldn't work on Mac as is?

As far as I know it's possible but the shared memory transport needs a special implementation for each platform, it isn't used very often in the first place, and no one has payed or created a PR for a macOS implementation.

**Comment by mitza-oci** [comment]
OpenDDS is using (via ACE) unnamed shared semaphores, which are in POSIX but not in Apple macOS (and ACE doesn't emulate):
https://stackoverflow.com/questions/1413785/sem-init-on-os-x

If this isn't already listed somewhere in the DevGuide or opendds.org we can add it.  It could be added to the FAQ or the OS section of https://opendds.org/documents/building.html.  The opendds.org site sources are in GitHub so you can contribute changes.

**Comment by mouse07410** [comment]
> If this isn't already listed somewhere in the DevGuide or opendds.org we can add it...

A few comments on the docs. Probably warrants a separate issue, but...

1. It would be nice if the relevant docs were included in what I pull from the GitHub. At least the `OpenDDS-latest.pdf`. 
2. It would be nice if the `README.md` or (maybe even better) `INSTALL.md` ad some more details on the configuration and testing. I personally think that one shouldn't expect a user to go through all the documents available on the Web site with a fine-tooth comb before trying to "just" download, build, and see if it passes most of the tests as-is (aka, if it builds/runs on the given system). And only after that actually starting dig deeper into the docs, and writing his/her own applications.
3. Yes, I think the `SHMEM` issue deserves being mentioned in the FAQ (as well as in the OS section of the Building page).
4. I'd like to add support for Wireshark to dissect traffic. I did not compile Wireshark from the source, so here's what I have libraries-wise:
```
$ find /Applications/Wireshark.app -name '*.dylib' -print
/Applications/Wireshark.app/Contents/PlugIns/platforminputcontexts/libqtvirtualkeyboardplugin.dylib
/Applications/Wireshark.app/Contents/PlugIns/mediaservice/libqavfmediaplayer.dylib
/Applications/Wireshark.app/Contents/PlugIns/mediaservice/libqtmedia_audioengine.dylib
/Applications/Wireshark.app/Contents/PlugIns/mediaservice/libqavfcamera.dylib
/Applications/Wireshark.app/Contents/PlugIns/platforms/libqcocoa.dylib
/Applications/Wireshark.app/Contents/PlugIns/printsupport/libcocoaprintersupport.dylib
/Applications/Wireshark.app/Contents/PlugIns/styles/libqmacstyle.dylib
/Applications/Wireshark.app/Contents/PlugIns/audio/libqtaudio_coreaudio.dylib
/Applications/Wireshark.app/Contents/PlugIns/bearer/libqgenericbearer.dylib
/Applications/Wireshark.app/Contents/PlugIns/iconengines/libqsvgicon.dylib
/Applications/Wireshark.app/Contents/PlugIns/imageformats/libqgif.dylib
/Applications/Wireshark.app/Contents/PlugIns/imageformats/libqwbmp.dylib
/Applications/Wireshark.app/Contents/PlugIns/imageformats/libqwebp.dylib
/Applications/Wireshark.app/Contents/PlugIns/imageformats/libqico.dylib
/Applications/Wireshark.app/Contents/PlugIns/imageformats/libqmacheif.dylib
/Applications/Wireshark.app/Contents/PlugIns/imageformats/libqjpeg.dylib
/Applications/Wireshark.app/Contents/PlugIns/imageformats/libqtiff.dylib
/Applications/Wireshark.app/Contents/PlugIns/imageformats/libqsvg.dylib
/Applications/Wireshark.app/Contents/PlugIns/imageformats/libqicns.dylib
/Applications/Wireshark.app/Contents/PlugIns/imageformats/libqtga.dylib
/Applications/Wireshark.app/Contents/PlugIns/imageformats/libqmacjp2.dylib
/Applications/Wireshark.app/Contents/PlugIns/virtualkeyboard/libqtvirtualkeyboard_thai.dylib
/Applications/Wireshark.app/Contents/PlugIns/virtualkeyboard/libqtvirtualkeyboard_openwnn.dylib
/Applications/Wireshark.app/Contents/PlugIns/virtualkeyboard/libqtvirtualkeyboard_hangul.dylib
/Applications/Wireshark.app/Contents/PlugIns/virtualkeyboard/libqtvirtualkeyboard_pinyin.dylib
/Applications/Wireshark.app/Contents/PlugIns/virtualkeyboard/libqtvirtualkeyboard_tcime.dylib
/Applications/Wireshark.app/Contents/Frameworks/libcares.2.dylib.dSYM/Contents/Resources/DWARF/libcares.2.dylib
/Applications/Wireshark.app/Contents/Frameworks/libspandsp.2.dylib
/Applications/Wireshark.app/Contents/Frameworks/libminizip.1.dylib
/Applications/Wireshark.app/Contents/Frameworks/libzstd.1.dylib.dSYM/Contents/Resources/DWARF/libzstd.1.dylib
/Applications/Wireshark.app/Contents/Frameworks/liblz4.1.dylib.dSYM/Contents/Resources/DWARF/liblz4.1.dylib
/Applications/Wireshark.app/Contents/Frameworks/libxml2.2.dylib
/Applications/Wireshark.app/Contents/Frameworks/libwsutil.11.dylib.dSYM/Contents/Resources/DWARF/libwsutil.11.dylib
/Applications/Wireshark.app/Contents/Frameworks/libbrotlidec.1.dylib
/Applications/Wireshark.app/Contents/Frameworks/libsnappy.1.dylib.dSYM/Contents/Resources/DWARF/libsnappy.1.dylib
/Applications/Wireshark.app/Contents/Frameworks/libwiretap.10.dylib.dSYM/Contents/Resources/DWARF/libwiretap.10.dylib
/Applications/Wireshark.app/Contents/Frameworks/libgnutls.30.dylib
/Applications/Wireshark.app/Contents/Frameworks/libgcrypt.20.dylib.dSYM/Contents/Resources/DWARF/libgcrypt.20.dylib
/Applications/Wireshark.app/Contents/Frameworks/libbrotlicommon.1.dylib
/Applications/Wireshark.app/Contents/Frameworks/libtiff.3.dylib.dSYM/Contents/Resources/DWARF/libtiff.3.dylib
/Applications/Wireshark.app/Contents/Frameworks/libgthread-2.0.0.dylib.dSYM/Contents/Resources/DWARF/libgthread-2.0.0.dylib
/Applications/Wireshark.app/Contents/Frameworks/libmaxminddb.0.dylib
/Applications/Wireshark.app/Contents/Frameworks/libglib-2.0.0.dylib.dSYM/Contents/Resources/DWARF/libglib-2.0.0.dylib
/Applications/Wireshark.app/Contents/Frameworks/libglib-2.0.0.dylib
/Applications/Wireshark.app/Contents/Frameworks/liblzma.5.dylib.dSYM/Contents/Resources/DWARF/liblzma.5.dylib
/Applications/Wireshark.app/Contents/Frameworks/libwireshark.13.dylib.dSYM/Contents/Resources/DWARF/libwireshark.13.dylib
/Applications/Wireshark.app/Contents/Frameworks/libwiretap.10.dylib
/Applications/Wireshark.app/Contents/Frameworks/libcares.2.dylib
/Applications/Wireshark.app/Contents/Frameworks/libnettle.6.dylib
/Applications/Wireshark.app/Contents/Frameworks/libgmp.10.dylib.dSYM/Contents/Resources/DWARF/libgmp.10.dylib
/Applications/Wireshark.app/Contents/Frameworks/libgmodule-2.0.0.dylib
/Applications/Wireshark.app/Contents/Frameworks/libwireshark.13.dylib
/Applications/Wireshark.app/Contents/Frameworks/libhogweed.4.dylib
/Applications/Wireshark.app/Contents/Frameworks/libintl.8.dylib.dSYM/Contents/Resources/DWARF/libintl.8.dylib
/Applications/Wireshark.app/Contents/Frameworks/libsmi.2.dylib.dSYM/Contents/Resources/DWARF/libsmi.2.dylib
/Applications/Wireshark.app/Contents/Frameworks/libintl.8.dylib
/Applications/Wireshark.app/Contents/Frameworks/libtiff.3.dylib
/Applications/Wireshark.app/Contents/Frameworks/libbrotlidec.1.dylib.dSYM/Contents/Resources/DWARF/libbrotlidec.1.dylib
/Applications/Wireshark.app/Contents/Frameworks/libgpg-error.0.dylib
/Applications/Wireshark.app/Contents/Frameworks/libgnutls.30.dylib.dSYM/Contents/Resources/DWARF/libgnutls.30.dylib
/Applications/Wireshark.app/Contents/Frameworks/libnghttp2.14.dylib
/Applications/Wireshark.app/Contents/Frameworks/libxml2.2.dylib.dSYM/Contents/Resources/DWARF/libxml2.2.dylib
/Applications/Wireshark.app/Contents/Frameworks/libgmodule-2.0.0.dylib.dSYM/Contents/Resources/DWARF/libgmodule-2.0.0.dylib
/Applications/Wireshark.app/Contents/Frameworks/libnghttp2.14.dylib.dSYM/Contents/Resources/DWARF/libnghttp2.14.dylib
/Applications/Wireshark.app/Contents/Frameworks/libbcg729.0.dylib.dSYM/Contents/Resources/DWARF/libbcg729.0.dylib
/Applications/Wireshark.app/Contents/Frameworks/libhogweed.4.dylib.dSYM/Contents/Resources/DWARF/libhogweed.4.dylib
/Applications/Wireshark.app/Contents/Frameworks/libwsutil.11.dylib
/Applications/Wireshark.app/Contents/Frameworks/libgthread-2.0.0.dylib
/Applications/Wireshark.app/Contents/Frameworks/liblz4.1.dylib
/Applications/Wireshark.app/Contents/Frameworks/libgcrypt.20.dylib
/Applications/Wireshark.app/Contents/Frameworks/liblzma.5.dylib
/Applications/Wireshark.app/Contents/Frameworks/libzstd.1.dylib
/Applications/Wireshark.app/Contents/Frameworks/libsmi.2.dylib
/Applications/Wireshark.app/Contents/Frameworks/libnettle.6.dylib.dSYM/Contents/Resources/DWARF/libnettle.6.dylib
/Applications/Wireshark.app/Contents/Frameworks/libbcg729.0.dylib
/Applications/Wireshark.app/Contents/Frameworks/libsnappy.1.dylib
/Applications/Wireshark.app/Contents/Frameworks/libssh.4.dylib
/Applications/Wireshark.app/Contents/Frameworks/libmaxminddb.0.dylib.dSYM/Contents/Resources/DWARF/libmaxminddb.0.dylib
/Applications/Wireshark.app/Contents/Frameworks/libbrotlicommon.1.dylib.dSYM/Contents/Resources/DWARF/libbrotlicommon.1.dylib
/Applications/Wireshark.app/Contents/Frameworks/libgmp.10.dylib
/Applications/Wireshark.app/Contents/Frameworks/libssh.4.dylib.dSYM/Contents/Resources/DWARF/libssh.4.dylib
/Applications/Wireshark.app/Contents/Frameworks/libgpg-error.0.dylib.dSYM/Contents/Resources/DWARF/libgpg-error.0.dylib
/Applications/Wireshark.app/Contents/Frameworks/libminizip.1.dylib.dSYM/Contents/Resources/DWARF/libminizip.1.dylib
/Applications/Wireshark.app/Contents/Frameworks/libspandsp.2.dylib.dSYM/Contents/Resources/DWARF/libspandsp.2.dylib
```
Question - what to point the config script to, to help it find what it needs?

**Comment by mitza-oci** [comment]
> A few comments on the docs. Probably warrants a separate issue, but...
> 
>     1. It would be nice if the relevant docs were included in what I pull from the GitHub. At least the `OpenDDS-latest.pdf`.

I don't have a problem with that.

>     2. It would be nice if the `README.md` or (maybe even better) `INSTALL.md` ad some more details on the configuration and testing. I personally think that one shouldn't expect a user to go through all the documents available on the Web site with a fine-tooth comb before trying to "just" download, build, and see if it passes most of the tests as-is (aka, if it builds/runs on the given system). And only after that actually starting dig deeper into the docs, and writing his/her own applications.

OK, we will integrate pull requests that change these files.  This is a good time to do it as we prepare for 3.14.

>     3. Yes, I think the `SHMEM` issue deserves being mentioned in the FAQ (as well as in the OS section of the Building page).

These are both hosted in GitHub so we will similarly integrate pull requests that change them.

>     4. I'd like to add support for Wireshark to dissect traffic. I did not compile Wireshark from the source, so here's what I have libraries-wise:

Do you want to dissect `rtps_udp` traffic or OpenDDS-protocol traffic? For the former, it's built in to Wireshark.  For the latter, `tools/dissector/README.md` has detailed instructions, including for macOS.

**Comment by mouse07410** [comment]
> As a workaround, see if you can use `ACE_TAO/ACE/bin/auto_run_tests.pl` instead. After sourcing `setenv.sh`, try something like:
> 
> ```
> $ACE_ROOT/bin/auto_run_tests.pl \
>   -d \
>   $DDS_ROOT/tests/security/security_tests.lst \
>   $DDS_ROOT/java/tests/dcps_java_tests.lst \
>   -Config RTPS -Config NO_SHMEM -Config DDS_NO_ORBSVCS -Config CXX11
> ```

I'm happy to report that running the above ran maybe 25-30% longer, and gave me 434 tests passing (as opposed to 258 with default), and 2 failing (as opposed to 11).

Thank you!

> `$ACE_ROOT/bin/auto_run_tests.pl` has a more helpful `--help` message too.

Yes it does.

```
. . . . .
Options:
    --help | -h              Display this help
    --ace | -a               Include the ACE tests
    --orb | -o               Include the TAO ORB tests
    --tao | -t               Include the TAO non-ORB tests
    --ciao-dance | -C        Include the CIAO and DAnCE tests
. . . . .
```

Which of these (`-a`, `-o`, `-t`) tests are built and can be run, after the doing the config/build command you showed?

Not that I really must exhaustively test ACE and TAO, but if tests of those are available - might as well exercise them, to add confidence of the build on my platform.

**Comment by iguessthislldo** [comment]
> Which of these (-a, -o, -t) tests are built and can be run, after the doing the config/build command you showed?

ACE and TAO tests aren't built by default when built using OpenDDS or ACE/TAO by itself. They are in `ACE_TAO/ACE/test` and `ACE_TAO/TAO/test`. After sourcing `setenv.sh`, you can build them manually by running `mwc.pl -type gnuace .` and `make` in both locations. Then you should be able to run them by passing `-aot` to `auto_run_tests.pl`.

**Comment by mouse07410** [comment]
I attempted to follow the above instructions - but got the following problem:
```
. . . . .
clang++ -std=c++11 -fdelayed-template-parsing -g -m64 -W -Wall -DACE_HAS_CUSTOM_EXPORT_MACROS=0   -I/Users/ur20980/src/OpenDDS/ACE_TAO/ACE -D__ACE_INLINE__ -I..  -c -o .obj/Compiler_Features_32_Test.o /Users/ur20980/src/OpenDDS/ACE_TAO/ACE/tests/Compiler_Features_32_Test.cpp
/Users/ur20980/src/OpenDDS/ACE_TAO/ACE/tests/Compiler_Features_32_Test.cpp:58:41: error: expected the class name after '~' to name a destructor
  this->u_.string_member_.std::string::~string ();
                                        ^
/Users/ur20980/src/OpenDDS/ACE_TAO/ACE/tests/Compiler_Features_32_Test.cpp:70:24: error: expected the class name after '~' to name a destructor
    u_.m.std::string::~string ();
                       ^
2 errors generated.
make[1]: *** [.obj/Compiler_Features_32_Test.o] Error 1
```

This is macOS Mojave 10.14.6, Xcode-11.3.1.

**Comment by mitza-oci** [comment]
Update Compiler_Features_32_Test.cpp for your clang version.  I'm guessing it's a newer Apple-clang that ACE isn't ready for yet.

**Comment by jwillemsen** [comment]
See https://github.com/DOCGroup/ACE_TAO/blob/816fca070860bbbb003aa59bc2acc69f98000cc0/ACE/tests/Compiler_Features_32_Test.cpp#L48 for the file/line to be updated, please update the clang version and open a new pull request to get the change integrated.

**Comment by mouse07410** [comment]
I'd rather not fork ACE and TAO. Here's the proposed patch:
```diff
diff --git a/ACE/tests/Compiler_Features_32_Test.cpp b/ACE/tests/Compiler_Features_32_Test.cpp
index a7374894..f51b9e16 100644
--- a/ACE/tests/Compiler_Features_32_Test.cpp
+++ b/ACE/tests/Compiler_Features_32_Test.cpp
@@ -45,7 +45,7 @@ A::u_type_::~u_type_ ()
 void A::clear ()
 {
 #if defined __clang__ && \
-    (defined __apple_build_version__ && __apple_build_version__ < 9100000 \
+    (defined __apple_build_version__ && __apple_build_version__ <= 11000033 \
      || __clang_major__ <= 9)
 #define CLANG_WORKAROUND
 #endif
```

**Comment by mitza-oci** [comment]
> I'd rather not fork ACE and TAO. Here's the proposed patch:
> 
> ```diff
> diff --git a/ACE/tests/Compiler_Features_32_Test.cpp b/ACE/tests/Compiler_Features_32_Test.cpp
> index a7374894..f51b9e16 100644
> --- a/ACE/tests/Compiler_Features_32_Test.cpp
> +++ b/ACE/tests/Compiler_Features_32_Test.cpp
> @@ -45,7 +45,7 @@ A::u_type_::~u_type_ ()
>  void A::clear ()
>  {
>  #if defined __clang__ && \
> -    (defined __apple_build_version__ && __apple_build_version__ < 9100000 \
> +    (defined __apple_build_version__ && __apple_build_version__ <= 11000033 \
>       || __clang_major__ <= 9)
>  #define CLANG_WORKAROUND
>  #endif
> ```

Please open an issue in the DOCGroup/ACE_TAO GitHub repository

**Comment by mouse07410** [comment]
Done: https://github.com/DOCGroup/ACE_TAO/issues/1042
